[28기 박준영] 메인페이지 트렌드 부분 구현 및 스타일 수정

joonyg10 opened the pull request:
## :: 최근 작업 주제 (하나 이상의 주제를 선택해주세요.)
- [X] 기능 추가
- [] 리뷰 반영
- [X] 리팩토링
- [] 버그 수정
- [] 컨벤션 수정

<br />

## :: 구현 목표 (해당 브랜치에서 구현하고자 하는 하나의 목표를 설정합니다.)
- https://www.samsung.com/sec/ 의 메인페이지를 클론 

<br />

## :: 구현 사항 설명 (작업한 내용을 상세하게 기록합니다.)

**0. 각 컴포넌트에 필요한 상수 데이터 저장하기 위한 data.js 생성**

**1. 메인 페이지 최상단의 이미지 슬라이더 추가**
 - 필요한 이미지는 public/main/hero에 추가해 img태그의 src에 경로 작성
 - 이미지는 컴포넌트 내에서 인덱스로 관리하도록 useState 적용
 - 양쪽 끝에 화살표 2개를 배치, 클릭시 다음 인덱스에 해당하는 이미지가 보이도록 구현
 - 현재 인덱스 (carouselIndex)를 css 변수로 stlye객체에 전달하였고 css에서 이를 이용해 transform에 적용
 - 이미지 밑의 navigator를 생성 
 - navigator 클릭 시 해당 인덱스로 이동토록 적용
 - navigator 오른쪽에는 재생, 정지 상태를 나타내는 버튼을 구현하였고 현재 재생상태인지는 isPlaying이란 state로 관리
 - isPlaying이 true인 경우 정지 버튼이 나타나며 setTimeOut을 활용하여 2.5초 이후에 자동으로 다음 인덱스로 넘어가도록 구현
 - 이 때, 컴포넌트 외부에 timer라는 변수를 선언하여 setTimeout에 의해 반환되는 ID를 저장
 - 만약 화살표 또는 navigator를 클릭하여 setTimeOut이 종료되기 전에 인덱스가 변경되는 경우 clearTimeOut으로 setTimeOut을 초기화 
 - 추가 구현 예정 : 슬라이더가 맨 앞, 뒤의 이미지를 가리키는 상황에서 화살표를 클릭 시 transform이 초기화 되지 않고 다음 페이지로 넘기듯 자연스럽게 넘어가는 효과 구현 예정

**2. Service 구현**
 - 이미지 슬라이더 하단에 서비스 종류를 보여주는 부분 구현
 - 탭은 있으나 작동하지 않으며 다른 화면이 보여지지는 않는다,
 - Service에서 hard coding으로 Icon, title, description등을 컴포넌트 외부에서 const CARDS라는 변수에 작성하였음
 - CARDS에 대해 map 메소드를 사용, Icon, title, description을 ServiceCard 컴포넌트의 prop으로 넘겨 주어 UI 구성
 - 각 Card에 대해 hover시 radial-gradient로 주어진 스타일 나타나도록 스타일 추가

**3. 상품 Preview에 대한 이미지 탭 구현**
 - 제품 미리보기에 관련된 이미지 슬라이더 추가
 - 1)과는 다르게 화살표는 제거 하였고 tab만을 통해 transform되도록 구현
 - 이미지의 높이를 105vh로 주었고 위의 탭 부분과 아래의 짧은 설명이 이미지 내부에 위치하도록 position: absolute로 위치
 - 짧은 설명의 p 태그 내부의 내용은 background: linear-gradient( )와 background-clip: text를 사용해 그라데이션을 텍스트에 적용

**4. Story 부분 구현**
- 메인 페이지에 story 부분 추가
- 오른쪽 탭에 따라 연동된 이미지가 가운데에 나타나도록 구현
- 초기에는 탭에 대한 index에 따라 img의 src로 index 번째 이미지를 불러오도록 하였으나 매끄럽게 넘어가지 않았음
- 모든 이미지를 초기에 불러온 뒤 position: absolute를 주었고 해당 index이외의 이미지에 z-index: -1을 주어 조금 더 매끄럽게 되도록 구현

**5. 트렌드 부분 구현**
- 메인페이지 마지막 sectio인 트렌드 부분 추가
- grid 레이아웃을 사용하였으며 각 Cell에 들어가는 부분도 중복 되었기에 TrendCard 컴포넌트 생성
- 이미지의 너비는 각 Cell의 너비로 맞추었고 높이는 조정이 필요한 경우 각 cell에 대해 --height라는 변수를 만들어 지정하였고,   
 TrendCard.scss에서 height: var(--height, 100%)로 하여 높이 조정 


**4. Component 분리** 
 1) Carousel Component
   - MainCarousel, Preview component에 대해 이미지 슬라이더가 중복적으로 쓰이기에 컴포넌트로 분리
   - 해당 부분에 필요한 모든 이미지 자료에 대해 map 메소드를 시행할 필요가 있기에 props로 자료를 담은 배열 = tabs를 받아옴
   - 이미지의 상단/하단의 탭과 연동될 필요가 있으나 탭부분은 따로 분리하였고 children으로 받아오도록 하였음

 2)  Navigator Component
   - 1)에서 기술한 Carousel과 연동을 위한 탭 부분을 컴포넌트화
   - Carousel과 연동되어야 하므로 currIdx라는 props으로 부모 컴포넌트에서 index를 state로 관리 및 carousel과 연동
   - Carousel마다 위치가 다르기에 props로 top을 받아와 position에 대해 top의 위치를 변경
   - Main Carousel에서는 play, pause버튼이 필요하였고 이를 children으로 받아오게 구현

 3) Info Component
   - 타이틀과 내용을 표시하기 위한 작은 컴포넌트 
   - 단순 presentational 컴포넌트 이지만 반복되는 부분이 많다고 판단하여 분리
   - title, description등을 props로 받고 표시

 4) Button Component
   - 메인페이지 에서 쓰일 기본적인 버튼을 컴포넌트화 

<br />

## :: 성장 포인트 (해당 기능을 구현하며 고민했던 사항이나 새로 알게된 부분, 어려웠던 점 등을 작성합니다.)


<br />

## :: 기타 질문 및 특이 사항
- 현재 메인 페이지의 Carousel에서 사용할 이미지들은 일단 public에 넣고 진행중입니다만, 백엔드에서 받아오도록 변경을 하는게 좋을지 모르겠습니다.

joonyg10 committed:
[Add] 메인페이지 이미지 슬라이더 구현

joonyg10 committed:
[ADD] 메인 페이지의 이미지 탭 추가
[FIX] 메인 이미지 슬라이더에서 1번째 navigator 클릭시 이동하지 않았던 오류 수정

joonyg10 committed:
[FIX] 컴포넌트 명 변경  ServiceCard -> Preview
[FIX] main.js의 import 컨벤션 수정

joonyg10 committed:
[ADD] service Card 구현 추가

joonyg10 committed:
[FIX] MainCarousel의 정지/재생 으로 인한 index state업데이트 오류 수정
[ADD] story component 디렉토리/파일 추가

joonyg10 committed:
[Refactoring] MainCarousel의 setTimeout부분 리펙토링

we-dohyeon commented on src/pages/Main/MainCarousel/MainCarousel.scss:
```diff
@@ -0,0 +1,87 @@
+@use "../../../styles/color" as color;
+@use "../../../styles/mixin" as mixin;
+
+.hero {
+  position: relative;
+  overflow: hidden;
```
hero라는 클래스명에 이유가 있을까요?

joonyg10 replied:
검색을 해봤는데 hero라는게 웹페이지 상단에 위치하는 큰 이미지라고 합니다. 현재 메인 페에지에서 hero란 클래스명은 MainCarousel 컴포넌트를 가리키는 데 이게 메인 상단에 위치한 이미지 슬라이더라 의미가 일치하다고 판단하여 hero라는 클래스명을 주었습니다.

we-dohyeon commented on src/pages/Main/PreviewTab/PreviewTab.jsx:
```diff
@@ -0,0 +1,55 @@
+  "Galaxy Academy",
+  "갤럭시 스튜디오",
+];
+const SERVICE_IMG_INDEXES = Array(5).fill(0);
```
해당 부분은 SERVICE_IMG_INDEXES라는 상수를 만들 필요 없이 상수 TABS를 그대로 사용해도 되겠네요!

joonyg10 replied:
네 문제 없어 보입니다! 바로 수정하겠습니다.

joonyg10 committed:
ADD: Main의 Story 부분 추가

joonyg10 committed:
Fix: ClassName 컨벤션 수정

we-dohyeon commented on src/pages/Main/PreviewTab/PreviewTab.jsx:
```diff
@@ -0,0 +1,55 @@
+    <section className="preview">
+      <div className="preview-tabs">
+        <h2>모바일 & Tablet</h2>
+        <ul onClick={navigateToIndex}>
+          {TABS.map((tab, idx) => (
+            <li
+              key={idx}
+              data-idx={idx}
+              className={`tab ${currentIdx === idx ? "current" : ""}`}
+            >
+              {tab}
+            </li>
+          ))}
+        </ul>
```
ul에 온클릭 이벤트를 걸고 li에 data-idx 속성을 부여해주셨는데
li태그에 직접 온클릭 이벤트를 걸고 idx값을 인자로 넘겨주면 훨씬 간편하겠네요!

joonyg10 replied:
넵!! 수정 후 push하겠습니다!

we-dohyeon commented on src/pages/Main/PreviewTab/PreviewTab.jsx:
```diff
@@ -0,0 +1,55 @@
+          ))}
+        </ul>
+      </div>
+      <ul className="preview-images" style={{ "--currIdx": `${currentIdx}` }}>
```
오! 엄청난 아이디어네요!! 굿굿! 좋습니다~

we-dohyeon commented on src/pages/Main/Services/Services.jsx:
```diff
@@ -0,0 +1,107 @@
+            Icon={card.Icon}
+            bg={card.bg}
+            title={card.title}
+            description={card.desc}
```
컨벤션에 따라 Icon 또한 icon으로 변경해주세요!
또, card={card}와 같이 한 번에 보내는 방법도 있습니다!

joonyg10 replied:
음 Icon이 컴포넌트이고 ServiceCard 컴포넌트로 넘겨주어 < icon />으로 사용하는 것이 어색해서 Icon으로 주었습니다.
아니면 Icon 컴포넌트를 ServiceCard의 children으로 주는 방식으로 변경해 보면 괜찮을까요?

we-dohyeon replied:
밑에 관련한 리뷰들이 더 있으니 참고하셔서 어떤 방법이 가장 효율적일지 고민해보세요!

we-dohyeon commented on src/pages/Main/Services/Services.jsx:
```diff
@@ -0,0 +1,107 @@
+import { BsHeadset } from "react-icons/bs";
+import "./Services.scss";
+
+const SERVICES = [
+  "셀렉트샵",
+  "새로운 갤럭시",
+  "삼성 케어 플러스",
+  "특별한 혜택",
+  "공동구매",
+];
+
+const CARDS = [
+  {
+    id: 1,
+    Icon: FaRedhat,
+    bg: "rgb(59, 196, 163)",
+    title: "갤럭시 캠퍼스 스토어",
+    desc: "대학생, 대학원생과 대학 교원, 교직원을 위한 교육 할인 스토어",
+  },
+  {
+    id: 2,
```
해당 컴포넌트 뿐만 아니라 전체적으로 상수데이터의 사용이 많아 가독성을 해치는 경우도 있습니다.
Main 폴더 하위에 data.js 파일을 하나 만들어서 상수 데이터들을 한 곳에서 관리해보세요!
필요한 컴포넌트에서는 import 해서 사용해주시면 됩니다!

joonyg10 replied:
넵! 수정하겠습니다!!

we-dohyeon commented on src/pages/Main/Services/ServiceCard/ServiceCard.jsx:
```diff
@@ -0,0 +1,13 @@
+import "./ServiceCard.scss";
+
+const ServiceCard = ({ Icon, bg, title, description }) => {
```
전 리뷰와 이어지는 부분인데 card 정보들을 한 번에 props로 넘겨준다면 card만 받아온 후 구조분해 할당하고 사용해주시면 됩니다!

joonyg10 replied:
네 그렇게 바꾸어 보겠습니다! 리뷰 감사합니다.

we-dohyeon commented on src/pages/Main/Services/Services.jsx:
```diff
@@ -0,0 +1,107 @@
+    title: "아울렛",
+    desc: "인기가전을 특별한 가격에",
+  },
+  {
+    id: 5,
+    Icon: FaHome,
+    bg: "rgb(165, 195, 155)",
+    title: "홈 피트니스",
+    desc: "회원이라면 누구나 홈 피트니스 베이직 1달 체험권",
+  },
```
icon : `<FaHome size={60} style={} /> `와 같이 컴포넌트를 직접 넣어서 보내주는 방법도 있는데 어느 쪽이 더 보기에 깔끔하다고 생각하시나요?

joonyg10 replied:
리뷰 달아주신 수정사항 완료 및 push완료 했습니다.

we-dohyeon commented on src/pages/Main/Services/Services.scss:
```diff
@@ -0,0 +1,34 @@
+@use "../../../styles/color" as color;
+@use "../../../styles/mixin" as mixin;
```
모든 scss 파일에서 import 하고 있는데 사용할 때만 import 해주세요!

we-dohyeon commented on src/pages/Main/MainCarousel/MainCarousel.jsx:
```diff
@@ -0,0 +1,105 @@
+} from "react-icons/io";
+import "./MainCarousel.scss";
+
+const LAST_CAROUSEL_INDEX = 6;
+const NAVIGATOR = [
+  "BESPOKE큐커",
+  "Galaxy book",
+  "공동구매",
+  "갤럭시 Z 플립3",
+  "갤럭시 워치 4",
+  "Neo QLED",
+  "Team Samsung",
+];
+const IMGARR = Array(7).fill(0);
```
1. LAST_CAROUSEL_INDEX는 NAVIGATOR.length - 1이 되겠네요.
2. IMGARR 또한 7길이의 배열을 만들필요 없이 NAVIGATOR를 그대로 사용하면 되겠죠?

we-dohyeon commented on src/pages/Main/MainCarousel/MainCarousel.jsx:
```diff
@@ -0,0 +1,105 @@
+          >
+            {navigator}
+          </li>
+        ))}
```
여기도 data-idx 속성을 이용하기 보다는 li태그에 직접 이벤트를 걸어서 idx를 인자로 보내주는 방법으로 변경해보세요!

we-dohyeon reviewed (commented):
수고하셨습니다! 준영님!!
리뷰 반영해주세요~~
하루만에 정말 많이 만드셨네요!

joonyg10 committed:
Fix: Review요청 사항 반영
Add: 각 컴포넌트에 들어가는 상수데이터를 Main/data.js을 생성 및 내부에 작성

joonyg10 committed:
Fix: 불필요한 scss import 제거"

joonyg10 committed:
ADD: MainCarousel, preview등의 여러 부분에서 쓰이는 이미지 슬라이더와 네비게이터를 컴포넌트로 만들어 생성
Fix: 기존의 캐러셀 등을  위에서 생성한 Carousel, Navigator로 대체

we-dohyeon commented:
리뷰 잘 반영해주셨네요!! 추가적인 리뷰사항 없습니다! 계속 진행해주세요!

joonyg10 committed:
Add: Info, Button 컴포넌트 추가
previewTab에 Info와 Button 컴포넌트 적용

joonyg10 committed:
Add: 메인 페이지의 Trend, TrendCard  컴포넌트 추가가
Add: Info라는 간단하게 정보를 나타내는 컴포넌트 추가

joonyg10 committed:
Refactoring: scss 파일에서 불필요한 스타일 수정
Fix: Story의 오른쪽 탭에 hover시 background대신 색상이 변하도록 수정
Add: ServiceCard에 hover시 보이도록 스타일 추가

we-dohyeon commented:
준영님! 다음 PR에서는 성장포인트 기대하겠습니다!
너무 잘 하고 계신만큼 스스로를 되돌아보면 찾을 수 있는 부분이 많을 것 같아요!
